### DISCUSSION TITLE
[Guidance] Resource access authorization

### DISCUSSION THREAD

**Discussion by IndikaUdagedara** [ckan/ckan#7713] [General]
## CKAN version: 2.10

We have a requirement to enforce users to log-in to preview and download resources whilst keeping the datasets public i.e. Datasets must be publicly searchable and browsable but when a user goes to preview/download resources (of public or private Datasets), they should be redirected to the login page.

This should apply to API as well. E.g. `resource_show` should return 401 if user doesn't send a valid API key. 

Are you able to guide us to implement this using CKAN's plugins toolkit? We looked at `IAuthFunctions` but it seems to operate at `action` layer rather than `views` i.e. we can add a custom auth function to `resource_show` but resource view seems to be using other actions under the hood (e.g. `package_show`), which we don't want to put behind authorization.

**Comment by kowh-ai** [comment]
This was discussed in the Tech Team meeting today...consensus was this looks more like a feature request and potentially a fair bit of work. Also, this would sit better in 'Discussions' rather than 'Issues' so will move it there

**Comment by hongjie123** [comment]
Hello, we also want to add this feature at the moment. Have you implemented it?
We have a requirement to enforce users to log-in to preview and download resources whilst keeping the datasets public.
